core/vm/instructions.go

```diff
@@ -486,7 +486,13 @@ func opBlockhash(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) (
 	if num64 >= lower && num64 < upper {
 		// if Prague is active, read it from the history contract (EIP 2935).
 		if evm.chainRules.IsPrague {
-			num.SetBytes(getBlockHashFromContract(num64, evm.StateDB, evm.Accesses).Bytes())
+			blockHash, statelessGas := getBlockHashFromContract(num64, evm.StateDB, evm.Accesses)
+			if num64 != upper-1 && interpreter.evm.chainRules.IsEIP4762 {
```

Comment:
@gballet, note here the `num64 != upper-1`. The reason I added this is that the block will always have a witness included for the `currentBlockNum - 1`.

So in theory, we can avoid charging people that use that same block hash that was already included (for free) by the protocol. For older blocks, then simply charge as usual.

This is what makes sense to me, but:
- Want to get your opinion.
- We can discuss/surface in VIC.
- Eventually I can create a PR to EIP-4762 since this rule becomes somewhat intertwined with 2935.

Reply:
that makes sense, and yet I see several reasons not to do this:

 * This adds a corner case, not a big one, but still, so we would need to modify the EIP, which means that we need to discuss it on the VIC indeed.
 * The behavior, from a user point of view, is a bit strange because it's usually quite expensive to call this instructions, but sometimes it's really cheap
 * Yes the value is added to the witness at the start of the block, but a client who doesn't have caching properly implemented could reload the data each time this tx is called
 * I don't expect many txs to refer to the block just before, especially in a verkle context where the proof takes longer to build
 * if for whatever reason we decide to change it later, this will be a long check

None of them is strong enough by itself to justify not doing this... but it really feels like a corner case for something that has virtually no chance of being of interest.

Reply:
Well, I agree that those arguments you wrote don't sound convincing... but I'm fine with removing this gas optimization for now, and we can see later if more people want to push for it.